frozen-abi/src/abi_example.rs

```diff
@@ -329,13 +323,38 @@ impl<T: AbiExample> AbiExample for std::sync::Arc<T> {
     }
 }
 
+// When T is weakly owned by the likes of `std::{sync, rc}::Weak`s, we need to uphold the ownership
```

Comment:
Appreciate the explanation here.
I think this makes sense because if we try serializing a `Weak` it will upgrade, and always return `None` without this leakage.

However, I'm not super clear on why we need to have the implementations for `Weak` at all.
The struct I see it used, which has now become `AbiExample`, is `PinnedVec`; but on the `PinnedVec` we have:

```rust
    #[serde(skip)]
    recycler: Weak<RecyclerX<PinnedVec<T>>>,
```

so shouldn't this `Weak` be ignored in any serialization? Does that not include the digests we do?

Reply:
> However, I'm not super clear on why we need to have the implementations for Weak at all.
> ...

another good question. yeah, some unspoken nuances are lying here... ;) i did my best to answer that in patch: fd6c28d7587d925442839f8f3bab8fe131b3edfb

> I think this makes sense because if we try serializing a Weak it will upgrade, and always return None without this leakage.

yep, i noticed this bug in `Weak::example()` later after creating this pr and this commit specifically fixed: f5bdb30fe9d52baa95190a38e2f10252534a65d6

as you correctly pointed out, `Weak` isn't used for actual abi digesting currently. but for perfection, i did the extra effort by adding unit test as well, should we serialize `Weak` in distant future... ;)